utils/frame/try-runtime/cli/src/commands/follow_chain.rs

```diff
@@ -42,51 +46,177 @@ pub struct FollowChainCmd {
+	///
+	/// All fetched headers are stored in `self.fetched_headers`.
+	async fn fetch(&mut self) {
+		let last_finalized = match self.subscription.next_header().await {
+			Some(header) => header,
+			None => return,
+		};
+
+		self.fetched_headers.push_front(last_finalized.clone());
+
+		let current_height = last_finalized.number();
+		let parent_height = current_height.sub(One::one());
+		let last_height = self.last_returned.unwrap_or(parent_height);
+
+		let mut parent_hash = last_finalized.parent_hash().clone();
+		for _ in 0u32..(parent_height.saturating_sub(last_height).try_into().unwrap_or_default()) {
+			let parent_header = self.header_provider.get_header(parent_hash).await;
+			self.fetched_headers.push_front(parent_header.clone());
+			parent_hash = *parent_header.parent_hash();
+		}
+	}
+
+	/// Get the next finalized header.
+	pub async fn next(&mut self) -> Option<Block::Header> {
+		if self.fetched_headers.is_empty() {
+			self.fetch().await;
+		}
+
+		if let Some(header) = self.fetched_headers.pop_front() {
+			self.last_returned = Some(*header.number());
+			Some(header)
+		} else {
+			None
+		}
+	}
+}
+
 pub(crate) async fn follow_chain<Block, ExecDispatch>(
 	shared: SharedParams,
 	command: FollowChainCmd,
 	config: Configuration,
 ) -> sc_cli::Result<()>
 where
-	Block: BlockT<Hash = H256> + serde::de::DeserializeOwned,
+	Block: BlockT<Hash = H256> + DeserializeOwned,
 	Block::Hash: FromStr,
-	Block::Header: serde::de::DeserializeOwned,
+	Block::Header: DeserializeOwned,
 	<Block::Hash as FromStr>::Err: Debug,
 	NumberFor<Block>: FromStr,
 	<NumberFor<Block> as FromStr>::Err: Debug,
 	ExecDispatch: NativeExecutionDispatch + 'static,
 {
 	let mut maybe_state_ext = None;
-
-	let client = WsClientBuilder::default()
-		.connection_timeout(std::time::Duration::new(20, 0))
-		.max_notifs_per_subscription(1024)
-		.max_request_body_size(u32::MAX)
-		.build(&command.uri)
-		.await
-		.unwrap();
-
-	log::info!(target: LOG_TARGET, "subscribing to {:?} / {:?}", SUB, UN_SUB);
-	let mut subscription: Subscription<Block::Header> =
-		client.subscribe(SUB, None, UN_SUB).await.unwrap();
+	let (_client, subscription) = start_subscribing::<Block::Header>(&command.uri).await;
 
 	let (code_key, code) = extract_code(&config.chain_spec)?;
 	let executor = build_executor::<ExecDispatch>(&shared, &config);
 	let execution = shared.execution;
 
-	loop {
-		let header = match subscription.next().await {
-			Some(Ok(header)) => header,
-			None => {
-				log::warn!("subscription closed");
-				break
-			},
-			Some(Err(why)) => {
-				log::warn!("subscription returned error: {:?}. Probably decoding has failed.", why);
-				continue
-			},
-		};
+	let header_provider: RpcHeaderProvider<Block> =
+		RpcHeaderProvider { uri: command.uri.clone(), _phantom: PhantomData {} };
+	let mut finalized_headers: FinalizedHeaders<
+		Block,
+		RpcHeaderProvider<Block>,
+		Subscription<Block::Header>,
+	> = FinalizedHeaders::new(header_provider, subscription);
```

Comment:
@niklasad1 this can be an abstraction that can be moved to sub-xt perhaps?

Reply:
I reckon that it would be useful to have this abstraction in subxt indeed but this `library` is generic of `Block` which would be a bit hard to use in subxt as it generates its own version of the `Config/System trait` from the metadata.

We could investigate whether it's would be possible to use subxt to get rid of some of the boiler plate here.

Reply:
just saying that the `subscribe_finalized` may always skip blocks, and it will be a very common scenario for devs -- perhaps subxt should provide its own wrapper here.

Reply:
yeah gotcha, I opened https://github.com/paritytech/subxt/issues/630